<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Tidy GenomicRanges (from Michael Love) - 2&nbsp; Compute overlaps</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../pages/promoter-marks.html" rel="next">
<link href="../pages/join-is-an-overlap.html" rel="prev">
<link href="../inst/assets/favicon.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
</head>
<body class="nav-sidebar docked slimcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/compute-overlaps.html"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Compute overlaps</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">Tidy GenomicRanges (from Michael Love)</a> 
        <div class="sidebar-tools-main">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-git"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/js2264/BiocBook.GRanges/">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://js2264.github.io/BiocBook.GRanges/devel/">
            Browse version `devel`
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://js2264.github.io/BiocBook.GRanges/3.17/">
            Browse version `3.17`
            </a>
          </li>
      </ul>
</div>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/join-is-an-overlap.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Join is an overlap</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/compute-overlaps.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Compute overlaps</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/promoter-marks.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Promoter marks</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/bootstrap-overlap.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Bootstrap overlap</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/snp-position-in-peaks.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">SNP position in peaks</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/gene-plots.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Gene plots</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/rnaseq-eda.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">RNA-seq EDA</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/many-genomic-models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Many genomic models</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/isoform-analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Isoform analysis</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/tximeta-gene-range.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Gene ranges in tximeta</span></span></a>
  </div>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">
<span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Compute overlaps</span>
</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Objectives
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Compute overlaps and summary statistics between two sets of genomic ranges</li>
</ul>
</div>
</div>
<p>Suppose we want to compute the mean genomic extent (distance from left-most to right-most basepair) of genes overlapping a set of query ranges.</p>
<p>We move on from the “classroom example” by seeing how we compute overlaps when the features are in genomic space. We will use <em>GRanges</em> in the Bioconductor package <em>GenomicRanges</em> <span class="citation" data-cites="Lawrence2013">(<a href="#ref-Lawrence2013" role="doc-biblioref">Lawrence et al. 2013</a>)</span> to represent the features and <em>plyranges</em> <span class="citation" data-cites="Lee2019">(<a href="#ref-Lee2019" role="doc-biblioref">Lee, Cook, and Lawrence 2019</a>)</span> to compute the overlaps, similarly to how we used <em>dplyr</em> to compute the overlaps in the previous analysis. So data.frame is to <em>dplyr</em> as GRanges is to <em>plyranges</em>.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Lawrence2013" class="csl-entry" role="listitem">
Lawrence, Michael, Wolfgang Huber, Hervé Pagès, Patrick Aboyoun, Marc Carlson, Robert Gentleman, Martin T Morgan, and Vincent J Carey. 2013. <span>“Software for Computing and Annotating Genomic Ranges.”</span> <em>PLoS Comput. Biol.</em> 9 (8): e1003118.
</div><div id="ref-Lee2019" class="csl-entry" role="listitem">
Lee, Stuart, Dianne Cook, and Michael Lawrence. 2019. <span>“Plyranges: A Grammar of Genomic Data Transformation.”</span> <em>Genome Biol.</em> 20 (1): 4.
</div></div><div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">plyranges</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note the structure of the <em>GRanges</em> object. We can create a <em>GRanges</em> from a data.frame by specifying two of: <code>start</code>, <code>end</code>, or <code>width</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>    seqnames<span class="op">=</span><span class="st">"chr1"</span>,</span>
<span>    start<span class="op">=</span><span class="fl">1</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">34e6</span>,<span class="fl">36e6</span>,<span class="fl">36.6e6</span><span class="op">)</span>,</span>
<span>    width<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2e5</span>,<span class="fl">2e5</span>,<span class="fl">1e5</span><span class="op">)</span>,</span>
<span>    strand<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"-"</span>,<span class="st">"-"</span><span class="op">)</span>,</span>
<span>    range_id<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"foo"</span>,<span class="st">"bar"</span>,<span class="st">"boo"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">r</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-construct.html">as_granges</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span>
<span><span class="va">r</span></span>
<span><span class="co">##  GRanges object with 3 ranges and 1 metadata column:</span></span>
<span><span class="co">##        seqnames            ranges strand | range_id</span></span>
<span><span class="co">##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;</span></span>
<span><span class="co">##    [1]     chr1 34000001-34200000      + |      foo</span></span>
<span><span class="co">##    [2]     chr1 36000001-36200000      - |      bar</span></span>
<span><span class="co">##    [3]     chr1 36600001-36700000      - |      boo</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In case you haven’t seen this before, <em>GRanges</em> objects have specific functions to pull out information. See <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">?GRanges</a></code> for details.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">r</span><span class="op">)</span></span>
<span><span class="co">##  [1] 3</span></span>
<span><span class="fu">seqnames</span><span class="op">(</span><span class="va">r</span><span class="op">)</span></span>
<span><span class="co">##  factor-Rle of length 3 with 1 run</span></span>
<span><span class="co">##    Lengths:    3</span></span>
<span><span class="co">##    Values : chr1</span></span>
<span><span class="co">##  Levels(1): chr1</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/strand.html">strand</a></span><span class="op">(</span><span class="va">r</span><span class="op">)</span></span>
<span><span class="co">##  factor-Rle of length 3 with 2 runs</span></span>
<span><span class="co">##    Lengths: 1 2</span></span>
<span><span class="co">##    Values : + -</span></span>
<span><span class="co">##  Levels(3): + - *</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s use <em>plyranges</em> to find the genes that overlap a region of interest.</p>
<p>Typically, it is prefered to use Ensembl or GENCODE gene annotation, the latter of which can by obtained from <em>AnnotationHub</em>. Ensembl gene annotation can be easily manipulated with the <em>ensembldb</em> package <span class="citation" data-cites="rainer2019">(<a href="#ref-rainer2019" role="doc-biblioref">Rainer, Gatto, and Weichenberger 2019</a>)</span>. Provided is some example code in an unevaluated code chunk:</p>
<div class="no-row-height column-margin column-container"><div id="ref-rainer2019" class="csl-entry" role="listitem">
Rainer, Johannes, Laurent Gatto, and Christian X Weichenberger. 2019. <span>“<span class="nocase">ensembldb: an R package to create and use Ensembl-based annotation resources</span>.”</span> <em>Bioinformatics</em> 35 (17): 3151–53. <a href="https://doi.org/10.1093/bioinformatics/btz031">https://doi.org/10.1093/bioinformatics/btz031</a>.
</div></div><div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jorainer/ensembldb">ensembldb</a></span><span class="op">)</span></span>
<span><span class="va">edb</span> <span class="op">&lt;-</span> <span class="va">...</span> <span class="co"># obtain from AnnotationHub or from GTF file</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcripts.html">genes</a></span><span class="op">(</span><span class="va">edb</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here, we will work with a static TxDb that is distributed as an annotation package in Bioconductor. We use this TxDb because it is an older genome release (hg19) that matches some ranges we will work with later, but generally it is recommended to use a recent (and versioned) Ensembl or GENCODE gene annotation.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The same function can be used to extract the gene ranges:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcripts.html">genes</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="op">)</span></span>
<span><span class="co">##    403 genes were dropped because they have exons located on both</span></span>
<span><span class="co">##    strands of the same reference sequence or on more than one reference</span></span>
<span><span class="co">##    sequence, so cannot be represented by a single genomic range.</span></span>
<span><span class="co">##    Use 'single.strand.genes.only=FALSE' to get all the genes in a</span></span>
<span><span class="co">##    GRangesList object, or use suppressMessages() to suppress this</span></span>
<span><span class="co">##    message.</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu">keepStandardChromosomes</span><span class="op">(</span><span class="va">g</span>, pruning.mode<span class="op">=</span><span class="st">"coarse"</span><span class="op">)</span></span>
<span><span class="va">g</span></span>
<span><span class="co">##  GRanges object with 23033 ranges and 1 metadata column:</span></span>
<span><span class="co">##          seqnames              ranges strand |     gene_id</span></span>
<span><span class="co">##             &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span></span>
<span><span class="co">##        1    chr19   58858172-58874214      - |           1</span></span>
<span><span class="co">##       10     chr8   18248755-18258723      + |          10</span></span>
<span><span class="co">##      100    chr20   43248163-43280376      - |         100</span></span>
<span><span class="co">##     1000    chr18   25530930-25757445      - |        1000</span></span>
<span><span class="co">##    10000     chr1 243651535-244006886      - |       10000</span></span>
<span><span class="co">##      ...      ...                 ...    ... .         ...</span></span>
<span><span class="co">##     9991     chr9 114979995-115095944      - |        9991</span></span>
<span><span class="co">##     9992    chr21   35736323-35743440      + |        9992</span></span>
<span><span class="co">##     9993    chr22   19023795-19109967      - |        9993</span></span>
<span><span class="co">##     9994     chr6   90539619-90584155      + |        9994</span></span>
<span><span class="co">##     9997    chr22   50961997-50964905      - |        9997</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 25 sequences (1 circular) from hg19 genome</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we are ready to test for overlaps. A left join gives us all the overlaps for ranges on the left side (here <code>r</code>). If a range on the left has no overlaps it appears with <code>NA</code> for the metadata columns of the right side ranges. If a range on the left side has multiple overlaps with the right side, it will appear multiple times in the output.</p>
<p>Keeping reading below on how to deal with this, if it is desired to have statistics on per-range overlaps.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">r</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html">join_overlap_left</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 10 ranges and 2 metadata columns:</span></span>
<span><span class="co">##         seqnames            ranges strand | range_id     gene_id</span></span>
<span><span class="co">##            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt; &lt;character&gt;</span></span>
<span><span class="co">##     [1]     chr1 34000001-34200000      + |      foo      114784</span></span>
<span><span class="co">##     [2]     chr1 36000001-36200000      - |      bar      127703</span></span>
<span><span class="co">##     [3]     chr1 36000001-36200000      - |      bar       23154</span></span>
<span><span class="co">##     [4]     chr1 36000001-36200000      - |      bar      339488</span></span>
<span><span class="co">##     [5]     chr1 36000001-36200000      - |      bar        5690</span></span>
<span><span class="co">##     [6]     chr1 36000001-36200000      - |      bar       63967</span></span>
<span><span class="co">##     [7]     chr1 36000001-36200000      - |      bar       79932</span></span>
<span><span class="co">##     [8]     chr1 36600001-36700000      - |      boo       27095</span></span>
<span><span class="co">##     [9]     chr1 36600001-36700000      - |      boo       55700</span></span>
<span><span class="co">##    [10]     chr1 36600001-36700000      - |      boo        9967</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If we want to exclude the zero matches cases, we can use an inner join:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">r</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html">join_overlap_inner</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 10 ranges and 2 metadata columns:</span></span>
<span><span class="co">##         seqnames            ranges strand | range_id     gene_id</span></span>
<span><span class="co">##            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt; &lt;character&gt;</span></span>
<span><span class="co">##     [1]     chr1 34000001-34200000      + |      foo      114784</span></span>
<span><span class="co">##     [2]     chr1 36000001-36200000      - |      bar      127703</span></span>
<span><span class="co">##     [3]     chr1 36000001-36200000      - |      bar       23154</span></span>
<span><span class="co">##     [4]     chr1 36000001-36200000      - |      bar      339488</span></span>
<span><span class="co">##     [5]     chr1 36000001-36200000      - |      bar        5690</span></span>
<span><span class="co">##     [6]     chr1 36000001-36200000      - |      bar       63967</span></span>
<span><span class="co">##     [7]     chr1 36000001-36200000      - |      bar       79932</span></span>
<span><span class="co">##     [8]     chr1 36600001-36700000      - |      boo       27095</span></span>
<span><span class="co">##     [9]     chr1 36600001-36700000      - |      boo       55700</span></span>
<span><span class="co">##    [10]     chr1 36600001-36700000      - |      boo        9967</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>What about other types of overlap? We can specify a distance, called a “gap”, if we want to also find when ranges are <em>near</em> each other, up to a maximum allowed distance. We can also specify the minimum amount of overlapping basepairs. Every overlap function in plyranges has arguments <code>maxgap</code> and <code>minoverlap</code>. For example, to see if ranges are 50kb from each other, we would specify <code>maxgap=5e4</code>. If we want to know if ranges are 50kb from a particular endpoint of another set of ranges, for example TSS, we could perform the operations <code><a href="https://rdrr.io/pkg/plyranges/man/ranges-anchor.html">anchor_5p()</a></code> followed by <code>mutate(width=1)</code>, before overlapping the sets.</p>
</div>
</div>
<p>We can also perform summarization by columns either in the <code>r</code> or the <code>g</code> object:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">r</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html">join_overlap_inner</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">range_id</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span><span class="op">(</span>count<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/n.html">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  DataFrame with 3 rows and 2 columns</span></span>
<span><span class="co">##    range_id     count</span></span>
<span><span class="co">##    &lt;factor&gt; &lt;integer&gt;</span></span>
<span><span class="co">##  1      bar         6</span></span>
<span><span class="co">##  2      boo         3</span></span>
<span><span class="co">##  3      foo         1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This is giving us the same information as the following:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">r</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-count-overlaps.html">count_overlaps</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span></span>
<span><span class="co">##  [1] 1 6 3</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Which can be added to the range data with a <code>mutate</code> call:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">r</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>overlaps <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-count-overlaps.html">count_overlaps</a></span><span class="op">(</span><span class="va">.</span>, <span class="va">g</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 3 ranges and 2 metadata columns:</span></span>
<span><span class="co">##        seqnames            ranges strand | range_id  overlaps</span></span>
<span><span class="co">##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt; &lt;integer&gt;</span></span>
<span><span class="co">##    [1]     chr1 34000001-34200000      + |      foo         1</span></span>
<span><span class="co">##    [2]     chr1 36000001-36200000      - |      bar         6</span></span>
<span><span class="co">##    [3]     chr1 36600001-36700000      - |      boo         3</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Up to this point, we could have used R’s native pipe <code>|&gt;</code> in place of the <em>magrittr</em> pipe <code>|&gt;</code>. However the above line won’t work with R’s native pipe, even if we use the underscore placeholder <code>_</code> that goes with R’s native pipe, instead of the dot placeholder <code>.</code>. This is because R’s native pipe doesn’t allow the placeholder to be within a nested expression (<code>count_overlaps</code> called before <code>mutate</code>). In addition, R’s native pipe only allows use of the placeholder once; sometimes it is convenient to access the incoming object more than once.</p>
</div>
</div>
<p>If we don’t care about multiple overlaps, but just want a binary variable that records if there was one or more overlaps or not, we can ask if the count of overlaps is greater than 0:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">r</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>overlaps_any <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-count-overlaps.html">count_overlaps</a></span><span class="op">(</span><span class="va">.</span>, <span class="va">g</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 3 ranges and 2 metadata columns:</span></span>
<span><span class="co">##        seqnames            ranges strand | range_id overlaps_any</span></span>
<span><span class="co">##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;    &lt;logical&gt;</span></span>
<span><span class="co">##    [1]     chr1 34000001-34200000      + |      foo         TRUE</span></span>
<span><span class="co">##    [2]     chr1 36000001-36200000      - |      bar         TRUE</span></span>
<span><span class="co">##    [3]     chr1 36600001-36700000      - |      boo         TRUE</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If we want to keep the information about the gene ranges, we swap the order of the ranges in the command:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">g</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html">join_overlap_inner</a></span><span class="op">(</span><span class="va">r</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 10 ranges and 2 metadata columns:</span></span>
<span><span class="co">##           seqnames            ranges strand |     gene_id range_id</span></span>
<span><span class="co">##              &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;factor&gt;</span></span>
<span><span class="co">##    114784     chr1 33979609-34631443      - |      114784      foo</span></span>
<span><span class="co">##    127703     chr1 36179477-36184790      - |      127703      bar</span></span>
<span><span class="co">##     23154     chr1 36023393-36032380      + |       23154      bar</span></span>
<span><span class="co">##     27095     chr1 36602170-36621654      - |       27095      boo</span></span>
<span><span class="co">##    339488     chr1 36038971-36060927      + |      339488      bar</span></span>
<span><span class="co">##     55700     chr1 36621803-36646441      + |       55700      boo</span></span>
<span><span class="co">##      5690     chr1 36035413-36107445      - |        5690      bar</span></span>
<span><span class="co">##     63967     chr1 36197713-36235551      - |       63967      bar</span></span>
<span><span class="co">##     79932     chr1 35899091-36023551      - |       79932      bar</span></span>
<span><span class="co">##      9967     chr1 36690017-36770957      + |        9967      boo</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 25 sequences (1 circular) from hg19 genome</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If we want strand specific overlaps, we can add <code>_directed</code>:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">g</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html">join_overlap_inner_directed</a></span><span class="op">(</span><span class="va">r</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 5 ranges and 2 metadata columns:</span></span>
<span><span class="co">##           seqnames            ranges strand |     gene_id range_id</span></span>
<span><span class="co">##              &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;factor&gt;</span></span>
<span><span class="co">##    127703     chr1 36179477-36184790      - |      127703      bar</span></span>
<span><span class="co">##     27095     chr1 36602170-36621654      - |       27095      boo</span></span>
<span><span class="co">##      5690     chr1 36035413-36107445      - |        5690      bar</span></span>
<span><span class="co">##     63967     chr1 36197713-36235551      - |       63967      bar</span></span>
<span><span class="co">##     79932     chr1 35899091-36023551      - |       79932      bar</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 25 sequences (1 circular) from hg19 genome</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>By turning the join around, we have access to the genomic range information about the genes. Now we can compute, e.g.&nbsp;the average genomic extent of the genes (first base to last base), per overlapping range.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">g</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html">join_overlap_inner_directed</a></span><span class="op">(</span><span class="va">r</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">range_id</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span><span class="op">(</span>count<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/n.html">n</a></span><span class="op">(</span><span class="op">)</span>, mean_width<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mean.html">mean</a></span><span class="op">(</span><span class="va">width</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  DataFrame with 2 rows and 3 columns</span></span>
<span><span class="co">##    range_id     count mean_width</span></span>
<span><span class="co">##    &lt;factor&gt; &lt;integer&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">##  1      bar         4    59911.8</span></span>
<span><span class="co">##  2      boo         1    19485.0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>What about <code>"foo"</code>? We need to add a <code>complete()</code> call to account for the fact that we are missing those overlaps after the join. We need to call the function explicitly from the <em>tidyr</em> package but by not loading the package we can avoid some function name conflicts with <em>plyranges</em>. Also we need to convert to tibble (explanation follows).</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/">tibble</a></span><span class="op">)</span></span>
<span><span class="va">g</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html">join_overlap_inner_directed</a></span><span class="op">(</span><span class="va">r</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">range_id</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span><span class="op">(</span>count<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/n.html">n</a></span><span class="op">(</span><span class="op">)</span>, mean_width<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mean.html">mean</a></span><span class="op">(</span><span class="va">width</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/complete.html">complete</a></span><span class="op">(</span><span class="va">range_id</span>, fill<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>count<span class="op">=</span><span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  # A tibble: 3 × 3</span></span>
<span><span class="co">##    range_id count mean_width</span></span>
<span><span class="co">##    &lt;fct&gt;    &lt;int&gt;      &lt;dbl&gt;</span></span>
<span><span class="co">##  1 bar          4     59912.</span></span>
<span><span class="co">##  2 boo          1     19485 </span></span>
<span><span class="co">##  3 foo          0        NA</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Why did we have to convert to tibble before running <code>complete()</code>? This is because metadata columns of <em>GRanges</em> objects are in a format called <em>DataFrame</em>, on which the <em>tidyr</em> / <em>dplyr</em> functions don’t know how to operate.</p>
<p>To access metadata columns of a <em>GRanges</em> object, you can use any of these paradigms:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">mcols</span><span class="op">(</span><span class="va">r</span><span class="op">)</span></span>
<span><span class="co">##  DataFrame with 3 rows and 1 column</span></span>
<span><span class="co">##    range_id</span></span>
<span><span class="co">##    &lt;factor&gt;</span></span>
<span><span class="co">##  1      foo</span></span>
<span><span class="co">##  2      bar</span></span>
<span><span class="co">##  3      boo</span></span>
<span><span class="fu">mcols</span><span class="op">(</span><span class="va">r</span><span class="op">)</span><span class="op">$</span><span class="va">range_id</span></span>
<span><span class="co">##  [1] foo bar boo</span></span>
<span><span class="co">##  Levels: bar boo foo</span></span>
<span><span class="va">r</span><span class="op">$</span><span class="va">range_id</span> <span class="co"># this works also</span></span>
<span><span class="co">##  [1] foo bar boo</span></span>
<span><span class="co">##  Levels: bar boo foo</span></span>
<span><span class="fu">mcols</span><span class="op">(</span><span class="va">r</span><span class="op">)</span><span class="op">[[</span><span class="st">"range_id"</span><span class="op">]</span><span class="op">]</span> <span class="co"># for programmatic access</span></span>
<span><span class="co">##  [1] foo bar boo</span></span>
<span><span class="co">##  Levels: bar boo foo</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>But if you want to work on them in with <em>tidyr</em> / <em>dplyr</em>, you need to first convert to tibble (or data.frame):</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">mcols</span><span class="op">(</span><span class="va">r</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">##  # A tibble: 3 × 1</span></span>
<span><span class="co">##    range_id</span></span>
<span><span class="co">##    &lt;fct&gt;   </span></span>
<span><span class="co">##  1 foo     </span></span>
<span><span class="co">##  2 bar     </span></span>
<span><span class="co">##  3 boo</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Reduce instead of summarize: Above when we used <code>group_by</code> and <code>summarize</code> we lost the original range data. Another option, to preserve the range data, is to use the function <code>reduce_ranges</code> within groups that we define (which can be <code>_directed</code> or not).</p>
</div>
</div>
<p>If we want to preserve the range information for the <code>r</code> object, we can start with <code>r</code> and proceed to join, group by, and reduce within groups. For an example of <code>reduce_ranges</code> used in the context of a genomic data analysis see <span class="citation" data-cites="Lee2020">Lee, Lawrence, and Love (<a href="#ref-Lee2020" role="doc-biblioref">2020</a>)</span>. In order to compute on the gene widths, we have to add that as a metadata column within the join. To keep the no-gene-overlapping ranges in <code>r</code>, we can count when the gene ID is not <code>NA</code>.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Lee2020" class="csl-entry" role="listitem">
Lee, Stuart, Michael Lawrence, and Michael I Love. 2020. <span>“Fluent Genomics with Plyranges and Tximeta.”</span> <em>F1000Res.</em> 9 (February): 109.
</div></div><div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">r</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html">join_overlap_left_directed</a></span><span class="op">(</span><span class="va">g</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>gene_width<span class="op">=</span><span class="va">width</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">range_id</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-reduce.html">reduce_ranges</a></span><span class="op">(</span></span>
<span>        count<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">gene_id</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        mean_width<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mean.html">mean</a></span><span class="op">(</span><span class="va">gene_width</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="co">##  GRanges object with 3 ranges and 3 metadata columns:</span></span>
<span><span class="co">##        seqnames            ranges strand | range_id     count mean_width</span></span>
<span><span class="co">##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt; &lt;integer&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">##    [1]     chr1 34000001-34200000      * |      foo         0         NA</span></span>
<span><span class="co">##    [2]     chr1 36000001-36200000      * |      bar         4    59911.8</span></span>
<span><span class="co">##    [3]     chr1 36600001-36700000      * |      boo         1    19485.0</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Hopefully, you’ve seen that there are many routes to compute the types of statistics of interest. The best way to decide which one to use is to think first: what do I want the final output to look like, and do I need to keep track of non-overlapping ranges? This will help dictate the way you set up your code, whether a <code>join</code> or a <code>mutate</code> to just tally a column of overlaps, etc., and whether a <code>complete</code> call is needed to fill in missing levels at the end of the analysis.</p>




<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../pages/join-is-an-overlap.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Join is an overlap</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../pages/promoter-marks.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Promoter marks</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">This book was built with <a href="https://github.com/js2264/BiocBook/">BiocBook</a> with <span class="emoji" data-emoji="heart">❤️</span></div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>


</body></html>