<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Tidy GenomicRanges (from Michael Love) - 9&nbsp; Isoform analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../pages/many-genomic-models.html" rel="prev">
<link href="../inst/assets/favicon.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
</head>
<body class="nav-sidebar docked slimcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/isoform-analysis.html"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Isoform analysis</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">Tidy GenomicRanges (from Michael Love)</a> 
        <div class="sidebar-tools-main">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-git"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/js2264/BiocBook.GRanges/">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://js2264.github.io/BiocBook.GRanges/devel/">
            Browse version `devel`
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://js2264.github.io/BiocBook.GRanges/3.17/">
            Browse version `3.17`
            </a>
          </li>
      </ul>
</div>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/join-is-an-overlap.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Join is an overlap</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/compute-overlaps.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Compute overlaps</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/promoter-marks.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Promoter marks</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/bootstrap-overlap.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Bootstrap overlap</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/snp-position-in-peaks.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">SNP position in peaks</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/gene-plots.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Gene plots</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/rnaseq-eda.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">RNA-seq EDA</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/many-genomic-models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Many genomic models</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/isoform-analysis.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Isoform analysis</span></span></a>
  </div>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">
<span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Isoform analysis</span>
</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Objectives
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Compare the structure of isoforms within a gene using grouping and disjoin operations</li>
</ul>
</div>
</div>
<p>There are many packages in Bioconductor that allow for isoform-level analysis across samples (also called transcript-level analysis). Packages that facilitate differential transcript analysis include <em>DEXSeq</em> and <em>DRIMSeq</em> (demonstrated in the <em>rnaseqDTU</em> workflow), and newer packages <em>satuRn</em> and <em>fishpond</em>.</p>
<p>Once you’ve identified isoforms of interest within a gene, perhaps isoforms that switch in terms of their expression after cells are treated, one can use the <a href="https://github.com/kvittingseerup/IsoformSwitchAnalyzeR">IsoformSwitchAnalyzeR</a> Bioconductor package to visualize and analyze a set of isoform switches <span class="citation" data-cites="Vitting-Seerup2019">(<a href="#ref-Vitting-Seerup2019" role="doc-biblioref">Vitting-Seerup and Sandelin 2019</a>)</span>. For example, one can test the functional consequences of a set of isoform switches in terms of the gain or loss of protein domains, or splicing-centric changes (e.g. alternative 3’ or 5’ acceptor sites, alternative transcription start or ends sites, etc.) <em>IsoformSwitchAnalyzeR</em> is a multi-feature and mature package for this type of analysis, but we can perform some simpler within-gene isoform comparisons using <em>plyranges</em>, mostly for demonstration.</p>
<div class="no-row-height column-margin column-container"><div id="ref-Vitting-Seerup2019" class="csl-entry" role="listitem">
Vitting-Seerup, Kristoffer, and Albin Sandelin. 2019. <span>“<span>IsoformSwitchAnalyzeR</span>: Analysis of Changes in Genome-Wide Patterns of Alternative Splicing and Its Functional Consequences.”</span> <em>Bioinformatics</em> 35 (21): 4469–71.
</div></div><p>Here we will suppose that we have somehow identified isoforms of interest, and we want to compare these isoforms to other isoforms of the same gene. For simplicity, we will focus on one isoform per gene, for a particular set of interesting genes, just picking isoforms at random for genes on <code>chr1</code>. We could use <em>plyranges</em> to compare various metadata about isoforms or exons e.g.&nbsp;RNA-seq or ChIP-seq coverage, sequence content, etc. But here we will just compare isoforms alone by the interval definitions. Then to reformulate:</p>
<p>Specific objective 1: compare one isoform per gene to the others, in terms of the extent from TSS to TES. What makes this isoform distinct?</p>
<p>We will start again with the transcript database we’ve used before:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="op">)</span></span>
<span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span></span>
<span><span class="va">txp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcripts.html">transcripts</a></span><span class="op">(</span><span class="va">txdb</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For further operations it will be convenient to have the <code>tx_id</code> be a character variable, and we filter now to a set of genes of interest (here just picking those on <code>chr1</code>):</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">plyranges</span><span class="op">)</span></span>
<span><span class="va">txp</span> <span class="op">&lt;-</span> <span class="va">txp</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>tx_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">tx_id</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">seqnames</span> <span class="op">==</span> <span class="st">"chr1"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>It is sometimes useful to have other identifiers, such as the gene ID (Entrez)…</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">txp</span> <span class="op">&lt;-</span> <span class="va">txp</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span></span>
<span>        gene_id <span class="op">=</span> <span class="fu">mapIds</span><span class="op">(</span></span>
<span>            <span class="va">txdb</span>, keys<span class="op">=</span><span class="va">tx_id</span>,</span>
<span>            column<span class="op">=</span><span class="st">"GENEID"</span>, keytype<span class="op">=</span><span class="st">"TXID"</span></span>
<span>        <span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">gene_id</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>…and the gene symbol. For simplicity we will keep genes that have a non-NA symbol but this step is not necessary.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span><span class="va">txp</span> <span class="op">&lt;-</span> <span class="va">txp</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span></span>
<span>        symbol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html">mapIds</a></span><span class="op">(</span></span>
<span>            <span class="va">org.Hs.eg.db</span>, keys<span class="op">=</span><span class="va">gene_id</span>,</span>
<span>            column<span class="op">=</span><span class="st">"SYMBOL"</span>, keytype<span class="op">=</span><span class="st">"ENTREZID"</span></span>
<span>        <span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">symbol</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The following is one way to identify which isoforms belong to multi-isoform genes:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">txp</span> <span class="op">&lt;-</span> <span class="va">txp</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">gene_id</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>ntxp <span class="op">=</span> <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/n.html">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">ungroup</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can now filter to the multi-isoform genes:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">txp</span> <span class="op">&lt;-</span> <span class="va">txp</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">ntxp</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here we arbitrarily pick one isoform per gene, first by identifying those in a tibble…</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/">tibble</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">pick_one</span> <span class="op">&lt;-</span> <span class="va">txp</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">gene_id</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample.int</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/n.html">n</a></span><span class="op">(</span><span class="op">)</span>, size<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">tx_id</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>…then we can label these in our <code>txp</code> object. We will track these with an integer, 1 for the isoform of interest, and 0 for the others.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">txp</span> <span class="op">&lt;-</span> <span class="va">txp</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>the_one <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html">as.integer</a></span><span class="op">(</span><span class="va">tx_id</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html">%in%</a></span> <span class="va">pick_one</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To identify which “parts” of the TSS-to-TES interval belong to which isoform, we can use <code>disjoin_ranges</code>. This breaks up the ranges, here grouped by gene, into distinct parts, and labels those according to whatever metadata variables we specify. Here we specify to combine <code>tx_id</code> into a collapsed string, but we could also perform numeric operations, e.g.&nbsp;<code>min</code> or <code>mean</code> etc. And we can specify more than one new variable to be created during the <code>disjoin_ranges</code> operation. As with <code>reduce_ranges</code>, this operation can be <code>_directed</code> or not.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">txp</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">gene_id</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-disjoin.html">disjoin_ranges</a></span><span class="op">(</span>tx_ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html">paste</a></span><span class="op">(</span><span class="va">tx_id</span>,collapse<span class="op">=</span><span class="st">","</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 4438 ranges and 2 metadata columns:</span></span>
<span><span class="co">##           seqnames              ranges strand |     gene_id</span></span>
<span><span class="co">##              &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span></span>
<span><span class="co">##       [1]     chr1         11874-14409      * |   100287102</span></span>
<span><span class="co">##       [2]     chr1       367659-368597      * |      729759</span></span>
<span><span class="co">##       [3]     chr1       621096-622034      * |      729759</span></span>
<span><span class="co">##       [4]     chr1       762971-763177      * |      643837</span></span>
<span><span class="co">##       [5]     chr1       763178-778984      * |      643837</span></span>
<span><span class="co">##       ...      ...                 ...    ... .         ...</span></span>
<span><span class="co">##    [4434]     chr1 249144203-249150234      * |       55657</span></span>
<span><span class="co">##    [4435]     chr1 249150235-249152264      * |       55657</span></span>
<span><span class="co">##    [4436]     chr1 249152265-249152912      * |       55657</span></span>
<span><span class="co">##    [4437]     chr1 249152913-249153125      * |       55657</span></span>
<span><span class="co">##    [4438]     chr1 249153126-249153315      * |       55657</span></span>
<span><span class="co">##                           tx_ids</span></span>
<span><span class="co">##                      &lt;character&gt;</span></span>
<span><span class="co">##       [1]                  1,2,3</span></span>
<span><span class="co">##       [2]                     11</span></span>
<span><span class="co">##       [3]                   4099</span></span>
<span><span class="co">##       [4]         14,15,16,17,18</span></span>
<span><span class="co">##       [5]      14,15,16,17,18,19</span></span>
<span><span class="co">##       ...                    ...</span></span>
<span><span class="co">##    [4434] 7962,7963,7964,7965,..</span></span>
<span><span class="co">##    [4435] 7963,7964,7965,7966,..</span></span>
<span><span class="co">##    [4436]    7964,7965,7966,7967</span></span>
<span><span class="co">##    [4437]         7965,7966,7967</span></span>
<span><span class="co">##    [4438]              7966,7967</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 93 sequences (1 circular) from hg19 genome</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here we try to answer the specific objective, by labeling which parts belong exclusively to the isoform of interest by computing <code>min(the_one)</code> (try to convince yourself that this does in fact identify these intervals).</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">txp</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">gene_id</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-disjoin.html">disjoin_ranges</a></span><span class="op">(</span>the_one_parts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">the_one</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">the_one_parts</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 257 ranges and 2 metadata columns:</span></span>
<span><span class="co">##          seqnames              ranges strand |     gene_id the_one_parts</span></span>
<span><span class="co">##             &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;     &lt;integer&gt;</span></span>
<span><span class="co">##      [1]     chr1       621096-622034      * |      729759             1</span></span>
<span><span class="co">##      [2]     chr1     2407754-2411622      * |        9651             1</span></span>
<span><span class="co">##      [3]     chr1     2495189-2495267      * |        8764             1</span></span>
<span><span class="co">##      [4]     chr1     3689334-3689351      * |      388588             1</span></span>
<span><span class="co">##      [5]     chr1     6268367-6268677      * |      388591             1</span></span>
<span><span class="co">##      ...      ...                 ...    ... .         ...           ...</span></span>
<span><span class="co">##    [253]     chr1 243651535-243663020      * |       10000             1</span></span>
<span><span class="co">##    [254]     chr1 244571794-244586136      * |         159             1</span></span>
<span><span class="co">##    [255]     chr1 246580715-246670644      * |       64754             1</span></span>
<span><span class="co">##    [256]     chr1 247492918-247495045      * |       84838             1</span></span>
<span><span class="co">##    [257]     chr1 249119835-249120154      * |       80851             1</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 93 sequences (1 circular) from hg19 genome</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Do the parts identified make sense if we check one gene?</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">txp</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">gene_id</span> <span class="op">==</span> <span class="st">"9651"</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 6 ranges and 6 metadata columns:</span></span>
<span><span class="co">##        seqnames          ranges strand |       tx_id     tx_name     gene_id</span></span>
<span><span class="co">##           &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">##    [1]     chr1 2407754-2436964      + |         134  uc001aji.1        9651</span></span>
<span><span class="co">##    [2]     chr1 2411623-2436892      + |         135  uc010nyz.2        9651</span></span>
<span><span class="co">##    [3]     chr1 2411623-2436964      + |         136  uc009vle.1        9651</span></span>
<span><span class="co">##    [4]     chr1 2411623-2436965      + |         137  uc001ajj.1        9651</span></span>
<span><span class="co">##    [5]     chr1 2411623-2436969      + |         138  uc001ajk.1        9651</span></span>
<span><span class="co">##    [6]     chr1 2430183-2436964      + |         139  uc001ajl.1        9651</span></span>
<span><span class="co">##             symbol      ntxp   the_one</span></span>
<span><span class="co">##        &lt;character&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">##    [1]       PLCH2         6         1</span></span>
<span><span class="co">##    [2]       PLCH2         6         0</span></span>
<span><span class="co">##    [3]       PLCH2         6         0</span></span>
<span><span class="co">##    [4]       PLCH2         6         0</span></span>
<span><span class="co">##    [5]       PLCH2         6         0</span></span>
<span><span class="co">##    [6]       PLCH2         6         0</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 93 sequences (1 circular) from hg19 genome</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s pause and consider what we’ve answered so far. We asked, for a given isoform per gene, what parts (intervals) uniquely define that isoform, when we just consider TSS-to-TES extent (ignoring the exonic/intronic structure). We started here mostly for simplicity, but typically we care about <em>transcribed</em> sequence, so let’s repeat this task, now considering what exonic parts are unique to one isoform per gene.</p>
<p>Specific objective 2: compare one isoform per gene to the others, in terms of the exonic intervals. What makes this isoform distinct?</p>
<p>To start, we will need a list of the exons, grouped by transcript.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ebt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcriptsBy.html">exonsBy</a></span><span class="op">(</span><span class="va">txdb</span>, by<span class="op">=</span><span class="st">"tx"</span><span class="op">)</span></span>
<span><span class="va">ebt</span> <span class="op">&lt;-</span> <span class="va">ebt</span><span class="op">[</span><span class="va">txp</span><span class="op">$</span><span class="va">tx_id</span><span class="op">]</span> <span class="co"># subset to those txp/genes of interest</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here, we could have also used <code>bind_ranges</code> but I find that for very large lists of ranges, <code>unlist</code> is faster:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">exons</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html">unlist</a></span><span class="op">(</span><span class="va">ebt</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html">select</a></span><span class="op">(</span><span class="va">exon_id</span>, <span class="va">exon_rank</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>tx_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">ebt</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/lengths.html">lengths</a></span><span class="op">(</span><span class="va">ebt</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">exons</span></span>
<span><span class="co">##  GRanges object with 67595 ranges and 3 metadata columns:</span></span>
<span><span class="co">##         seqnames              ranges strand |   exon_id exon_rank       tx_id</span></span>
<span><span class="co">##            &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;integer&gt; &lt;character&gt;</span></span>
<span><span class="co">##       1     chr1         11874-12227      + |         1         1           1</span></span>
<span><span class="co">##       1     chr1         12613-12721      + |         3         2           1</span></span>
<span><span class="co">##       1     chr1         13221-14409      + |         5         3           1</span></span>
<span><span class="co">##       2     chr1         11874-12227      + |         1         1           2</span></span>
<span><span class="co">##       2     chr1         12595-12721      + |         2         2           2</span></span>
<span><span class="co">##     ...      ...                 ...    ... .       ...       ...         ...</span></span>
<span><span class="co">##    7967     chr1 249149757-249149834      - |     27460         7        7967</span></span>
<span><span class="co">##    7967     chr1 249149579-249149657      - |     27459         8        7967</span></span>
<span><span class="co">##    7967     chr1 249148136-249148250      - |     27458         9        7967</span></span>
<span><span class="co">##    7967     chr1 249144881-249144980      - |     27457        10        7967</span></span>
<span><span class="co">##    7967     chr1 249144203-249144715      - |     27456        11        7967</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 93 sequences (1 circular) from hg19 genome</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>exons</code> ranges are missing some of our key metadata from <code>txp</code>. We can add this, by first <code>select</code>-ing what we want from <code>txp</code>, turning this into a tibble and <code>left_join</code>-ing to the <code>exons</code>. I add an <code>all.equal</code> step to make sure we have the two tables lined up, before we add the extra columns with <code>cbind</code>.</p>
<p>Some of this code wouldn’t be necessary for TranscriptDb with more details <code>exons</code> output, as with <em>ensembldb</em>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">txp_data</span> <span class="op">&lt;-</span> <span class="va">txp</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html">select</a></span><span class="op">(</span><span class="va">tx_id</span>, <span class="va">gene_id</span>, <span class="va">ntxp</span>, <span class="va">the_one</span>, .drop_ranges<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">ids</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>tx_id <span class="op">=</span> <span class="va">exons</span><span class="op">$</span><span class="va">tx_id</span><span class="op">)</span>,</span>
<span>    <span class="va">txp_data</span>,</span>
<span>    by<span class="op">=</span><span class="st">"tx_id"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">ids</span></span>
<span><span class="co">##  # A tibble: 67,595 × 4</span></span>
<span><span class="co">##    tx_id gene_id    ntxp the_one</span></span>
<span><span class="co">##    &lt;chr&gt; &lt;chr&gt;     &lt;int&gt;   &lt;int&gt;</span></span>
<span><span class="co">##  1 1     100287102     3       1</span></span>
<span><span class="co">##  2 1     100287102     3       1</span></span>
<span><span class="co">##  3 1     100287102     3       1</span></span>
<span><span class="co">##  4 2     100287102     3       0</span></span>
<span><span class="co">##  5 2     100287102     3       0</span></span>
<span><span class="co">##  6 2     100287102     3       0</span></span>
<span><span class="co">##  # ℹ 67,589 more rows</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.equal.html">all.equal</a></span><span class="op">(</span><span class="va">exons</span><span class="op">$</span><span class="va">tx_id</span>, <span class="va">ids</span><span class="op">$</span><span class="va">tx_id</span><span class="op">)</span></span>
<span><span class="co">##  [1] TRUE</span></span>
<span><span class="fu">mcols</span><span class="op">(</span><span class="va">exons</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html">cbind</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">exons</span><span class="op">)</span>, <span class="va">ids</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">tx_id</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">exons</span></span>
<span><span class="co">##  GRanges object with 67595 ranges and 6 metadata columns:</span></span>
<span><span class="co">##         seqnames              ranges strand |   exon_id exon_rank       tx_id</span></span>
<span><span class="co">##            &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;integer&gt; &lt;character&gt;</span></span>
<span><span class="co">##       1     chr1         11874-12227      + |         1         1           1</span></span>
<span><span class="co">##       1     chr1         12613-12721      + |         3         2           1</span></span>
<span><span class="co">##       1     chr1         13221-14409      + |         5         3           1</span></span>
<span><span class="co">##       2     chr1         11874-12227      + |         1         1           2</span></span>
<span><span class="co">##       2     chr1         12595-12721      + |         2         2           2</span></span>
<span><span class="co">##     ...      ...                 ...    ... .       ...       ...         ...</span></span>
<span><span class="co">##    7967     chr1 249149757-249149834      - |     27460         7        7967</span></span>
<span><span class="co">##    7967     chr1 249149579-249149657      - |     27459         8        7967</span></span>
<span><span class="co">##    7967     chr1 249148136-249148250      - |     27458         9        7967</span></span>
<span><span class="co">##    7967     chr1 249144881-249144980      - |     27457        10        7967</span></span>
<span><span class="co">##    7967     chr1 249144203-249144715      - |     27456        11        7967</span></span>
<span><span class="co">##             gene_id      ntxp   the_one</span></span>
<span><span class="co">##         &lt;character&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">##       1   100287102         3         1</span></span>
<span><span class="co">##       1   100287102         3         1</span></span>
<span><span class="co">##       1   100287102         3         1</span></span>
<span><span class="co">##       2   100287102         3         0</span></span>
<span><span class="co">##       2   100287102         3         0</span></span>
<span><span class="co">##     ...         ...       ...       ...</span></span>
<span><span class="co">##    7967       55657         6         0</span></span>
<span><span class="co">##    7967       55657         6         0</span></span>
<span><span class="co">##    7967       55657         6         0</span></span>
<span><span class="co">##    7967       55657         6         0</span></span>
<span><span class="co">##    7967       55657         6         0</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 93 sequences (1 circular) from hg19 genome</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We repeat similar code as performed above with <code>txp</code>, now identifying parts of exons that are unique to the isoform of interest, per gene:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">exon_parts</span> <span class="op">&lt;-</span> <span class="va">exons</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">gene_id</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-disjoin.html">disjoin_ranges</a></span><span class="op">(</span>the_one_parts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">the_one</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">the_one_parts</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">exon_parts</span></span>
<span><span class="co">##  GRanges object with 1604 ranges and 2 metadata columns:</span></span>
<span><span class="co">##           seqnames              ranges strand |     gene_id the_one_parts</span></span>
<span><span class="co">##              &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;     &lt;integer&gt;</span></span>
<span><span class="co">##       [1]     chr1       621096-622034      * |      729759             1</span></span>
<span><span class="co">##       [2]     chr1       776580-778984      * |      643837             1</span></span>
<span><span class="co">##       [3]     chr1       898412-898488      * |      339451             1</span></span>
<span><span class="co">##       [4]     chr1       898634-898716      * |      339451             1</span></span>
<span><span class="co">##       [5]     chr1       898885-899229      * |      339451             1</span></span>
<span><span class="co">##       ...      ...                 ...    ... .         ...           ...</span></span>
<span><span class="co">##    [1600]     chr1 247301447-247302017      * |        7678             1</span></span>
<span><span class="co">##    [1601]     chr1 247493268-247493383      * |       84838             1</span></span>
<span><span class="co">##    [1602]     chr1 247494734-247495045      * |       84838             1</span></span>
<span><span class="co">##    [1603]     chr1 249120034-249120154      * |       80851             1</span></span>
<span><span class="co">##    [1604]     chr1 249152521-249152710      * |       55657             1</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 93 sequences (1 circular) from hg19 genome</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To confirm that we’ve identified the right parts, let’s use <em>plotgardener</em> to visualize a particular gene:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">txp</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">gene_id</span> <span class="op">==</span> <span class="st">"339451"</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 5 ranges and 6 metadata columns:</span></span>
<span><span class="co">##        seqnames        ranges strand |       tx_id     tx_name     gene_id</span></span>
<span><span class="co">##           &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">##    [1]     chr1 895967-901099      + |          55  uc001aca.2      339451</span></span>
<span><span class="co">##    [2]     chr1 896829-897858      + |          56  uc001acb.1      339451</span></span>
<span><span class="co">##    [3]     chr1 897009-897858      + |          57  uc010nya.1      339451</span></span>
<span><span class="co">##    [4]     chr1 897461-901099      + |          58  uc001acc.2      339451</span></span>
<span><span class="co">##    [5]     chr1 897735-899229      + |          59  uc010nyb.1      339451</span></span>
<span><span class="co">##             symbol      ntxp   the_one</span></span>
<span><span class="co">##        &lt;character&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">##    [1]      KLHL17         5         0</span></span>
<span><span class="co">##    [2]      KLHL17         5         0</span></span>
<span><span class="co">##    [3]      KLHL17         5         0</span></span>
<span><span class="co">##    [4]      KLHL17         5         0</span></span>
<span><span class="co">##    [5]      KLHL17         5         1</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 93 sequences (1 circular) from hg19 genome</span></span>
<span><span class="va">tx_to_show</span> <span class="op">&lt;-</span> <span class="va">txp</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">gene_id</span> <span class="op">==</span> <span class="st">"339451"</span> <span class="op">&amp;</span> <span class="va">the_one</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">tx_name</span><span class="op">)</span></span>
<span><span class="va">these_parts</span> <span class="op">&lt;-</span> <span class="va">exon_parts</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">gene_id</span> <span class="op">==</span> <span class="st">"339451"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We lay out a page zooming into this gene and its isoforms:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://phanstiellab.github.io/plotgardener">plotgardener</a></span><span class="op">)</span></span>
<span><span class="va">par</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/pgParams.html">pgParams</a></span><span class="op">(</span></span>
<span>    chrom <span class="op">=</span> <span class="st">"chr1"</span>, </span>
<span>    chromstart <span class="op">=</span> <span class="fl">895.9e3</span>, chromend <span class="op">=</span> <span class="fl">901.2e3</span>,</span>
<span>    assembly <span class="op">=</span> <span class="st">"hg19"</span>, just <span class="op">=</span> <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We will highlight our isoform of interest:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hilite</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>transcript<span class="op">=</span><span class="va">tx_to_show</span>, color<span class="op">=</span><span class="st">"magenta"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Finally, putting it all together:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/pageCreate.html">pageCreate</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">5</span>, height <span class="op">=</span> <span class="fl">2.5</span>, showGuides <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/plotTranscripts.html">plotTranscripts</a></span><span class="op">(</span></span>
<span>    params <span class="op">=</span> <span class="va">par</span>, x <span class="op">=</span> <span class="fl">0.5</span>, y <span class="op">=</span> <span class="fl">1.5</span>, width <span class="op">=</span> <span class="fl">4</span>, height <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>    transcriptHighlights <span class="op">=</span> <span class="va">hilite</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/plotRanges.html">plotRanges</a></span><span class="op">(</span></span>
<span>    <span class="va">these_parts</span>, fill<span class="op">=</span><span class="st">"darkorchid"</span>,</span>
<span>    params <span class="op">=</span> <span class="va">par</span>, x <span class="op">=</span> <span class="fl">0.5</span>, y <span class="op">=</span> <span class="fl">1.75</span>, width <span class="op">=</span> <span class="fl">4</span>, height <span class="op">=</span> <span class="fl">.25</span></span>
<span><span class="op">)</span></span>
<span><span class="va">label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html">paste</a></span><span class="op">(</span><span class="st">"unique parts of"</span>, <span class="va">tx_to_show</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/plotText.html">plotText</a></span><span class="op">(</span></span>
<span>    label <span class="op">=</span> <span class="va">label</span>, fontcolor <span class="op">=</span> <span class="st">"darkorchid"</span>,</span>
<span>    params <span class="op">=</span> <span class="va">par</span>, x <span class="op">=</span> <span class="fl">3.1</span>, y <span class="op">=</span> <span class="fl">1.75</span>,</span>
<span>    just <span class="op">=</span> <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"bottom"</span><span class="op">)</span>, fontsize <span class="op">=</span> <span class="fl">8</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/plotGenomeLabel.html">plotGenomeLabel</a></span><span class="op">(</span></span>
<span>    params <span class="op">=</span> <span class="va">par</span>, x <span class="op">=</span> <span class="fl">0.5</span>, y <span class="op">=</span> <span class="fl">2</span>, length <span class="op">=</span> <span class="fl">4</span>,</span>
<span>    just <span class="op">=</span> <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"top"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="isoform-analysis_files/figure-html/isoform-plot-1.png" class="img-fluid figure-img" width="480"></p>
</figure>
</div>
</div>
</div>
<p>Questions:</p>
<ul>
<li>How else could we have found parts of one isoform per gene, that do not belong to any other isoforms of the genes. Would other approaches have any limitations?</li>
</ul>




<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../pages/many-genomic-models.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Many genomic models</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">This book was built with <a href="https://github.com/js2264/BiocBook/">BiocBook</a> with <span class="emoji" data-emoji="heart">❤️</span></div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>


</body></html>